File: contrib/mptmodels/mpt2python.py
from tulip import hybrid
import polytope
import scipy.io

"""
Contains functions that will read a .mat file exported by the MATLAB function
mpt2python and import it to either a PwaSysDyn or a LtiSysDyn.

Stephanie Tsuei, June 2014
"""


def load(filename):
    data = scipy.io.loadmat(filename)
    islti = bool(data['islti'][0][0])
    ispwa = bool(data['ispwa'][0][0])

    if islti:
        sys = load_lti(data['A'], data['B'], data['domainA'],
                       data['domainB'], data['UsetA'], data['UsetB'])

    elif ispwa:
        nlti = len(data['A'][0])
        lti_systems = []

        for i in xrange(nlti):
            A = data['A'][0][i]
            B = data['B'][0][i]
            K = data['K'][0][i]
            domainA = data['domainA'][0][i]
            domainB = data['domainB'][0][i]
            UsetA = data['UsetA'][0][i]
            UsetB = data['UsetB'][0][i]

            ltisys = load_lti(A, B, K, domainA, domainB, UsetA, UsetB)
            lti_systems.append(ltisys)

        cts_ss = polytope.Polytope(data['ctsA'], data['ctsB'])
        sys = hybrid.PwaSysDyn(list_subsys=lti_systems, domain=cts_ss)

    return sys



def load_lti(A, B, K, domainA, domainB, UsetA, UsetB):
    domain = polytope.Polytope(domainA, domainB)
    Uset = polytope.Polytope(UsetA, UsetB)

    lti = hybrid.LtiSysDyn(A=A, B=B, K=K, domain=domain, Uset=Uset)
    return lti

File: contrib/mptmodels/mpt2python.m
function mpt2python(model, filename)
% Takes a MPT system data structure (either LTISystem or PWASystem "model") and
% exports it to a .mat file "filename". This file can then be read back
% into Python to create a LtiSysDyn or PwaSysDyn in Tulip.

    if isa(model,'LTISystem')
        [A, B, K, domainA, domainB, UsetA, UsetB] = lti2python(model);
        islti = 1;
        ispwa = 0;
        save(filename, 'A', 'B', 'K', 'domainA', 'domainB', 'UsetA', ...
            'UsetB', 'islti', 'ispwa');
    elseif isa(model, 'PWASystem')
        [A, B, K, domainA, domainB, UsetA, UsetB, ctsA, ctsB] = ...
            pwa2python(model);
        islti = 0;
        ispwa = 1;
        save(filename, 'A', 'B', 'K', 'domainA', 'domainB', 'UsetA', ...
            'UsetB', 'ctsA', 'ctsB', 'islti', 'ispwa');
    end


    function [A,B,K,domainA,domainB,UsetA,UsetB,ctsA,ctsB] = pwa2python(pwasys)

        % Dynamics matrices already stored conveniently
        A = pwasys.A;
        B = pwasys.B;
        K = pwasys.f;

        % Get domain matrices for each of the lti systems
        domainA = cell(1,pwasys.ndyn);
        domainB = cell(1,pwasys.ndyn);
        UsetA = cell(1,pwasys.ndyn);
        UsetB = cell(1,pwasys.ndyn);
        cts_ss = cell(1,pwasys.ndyn);
        for i = 1:pwasys.ndyn
            ltisys = pwasys.toLTI(i);
            [~,~,~,dA,dB,uA,uB] = lti2python(ltisys);
            domainA{i} = dA;
            domainB{i} = dB;
            UsetA{i} = uA;
            UsetB{i} = uB;
            cts_ss{i} = Polyhedron(dA, dB);
        end

        % Get domain of the whole state
        cts_ss = PolyUnion('Set', [cts_ss{:}]);
        cts_ss.merge;
        ctsA = cts_ss.Set.A;
        ctsB = cts_ss.Set.b;
    end


    function [A,B,K,domainA,domainB,UsetA,UsetB] = lti2python(ltisys)

        % Make sure that the domain was set with 'xu' and not just 'x'
        if (ltisys.domain.Dim ~= (ltisys.nx + ltisys.nu))
            error('Domain of LTISystem must be set with xu option');
        end

        % Dynamics matrices
        A = ltisys.A;
        B = ltisys.B;
        K = ltisys.f;

        % State domain
        xdomain = projection(ltisys.domain, 1:1:ltisys.nx);
        domainA = xdomain.A;
        domainB = xdomain.b;

        % Input domain - rearrange order of cols so that domain is
        % [u;x] instead of [x;u]
        UsetA = ltisys.domain.A;
        UsetB = ltisys.domain.b;
        UsetA = [UsetA(:,ltisys.nx+1:end) UsetA(:,1:ltisys.nx)];
    end

end
